### DISCUSSION TITLE
Auth session add constraint (v4.1)

### DISCUSSION THREAD

**Discussion by Mzzay** [adonisjs/core#1667] [Help] [answered]
Hi, is it possible to add constraint on authentication ? 

I want to log the user just if he has a specific value in db.

Something like this : 
`
session: {
    serializer: 'lucid',
    model: 'App/Models/User',
    scheme: 'session',
    uid: 'email',
    password: 'password',
    typeId: 'typel_id'
},
`

and use like this : 
`await auth.attempt(email, password,typeId)`

Thanks for help.

**Comment by thetutlage** [comment, marked as answer]
Well, you can fetch the user yourself and then use `auth.login` method. For example:

```ts
const user = await User
  .query()
  .where('email', email)
  .where('other', 'constraints')
  .firstOrFail()


const isValid = await Hash.verify(user.password, password)
if(isValid) {
  await auth.login(user)
}
```

**Reply by Mzzay** [reply]
what's the difference between your code and this one : 

```js
const user = await User
  .query()
  .where('email', email)
  .where('other', 'constraints')
  .firstOrFail()

await auth.attempt(user.email,password)
```
The result is the same no ?

**Reply by thetutlage** [reply]
`auth.attempt` makes a database query. So potentially you are making two database queries

**Reply by Mzzay** [reply]
`isValid` is always `false` but `user` is not empty (user data)

Sure that `user.password` is the user's password ? 
```js
const user = await User
                    .query()
                    .where('email', email)
                    .where('otherparams', otherpam)
                    .firstOrFail()
              
 const isValid = await Hash.verify(user.password, password)
if(isValid) {
    await auth.login(userData)

    return response.send({ 
           success : true,
           message : "User connected",
           data : user
     })
}
```

**Reply by thetutlage** [reply]
Is your password hashed inside the database? Also can you please log the values `user.password` and `password` and share them?

**Reply by Mzzay** [reply]
yes in my db the password is hash, `user.password` is undefined and password is the password sent by the user not hashed. 

To access to the user's password I need to use `user.$attributes.password`

**Reply by thetutlage** [reply]
Why `user.password` is undefined? It should be the hashed value

**Reply by thetutlage** [reply]
Can you share your user model?

**Reply by Mzzay** [reply]
it's the default user model: 

```js
class User extends Model {
  static boot () {
    super.boot()

    /**
     * A hook to hash the user password before saving
     * it to the database.
     */
    this.addHook('beforeSave', async (userInstance) => {
      if (userInstance.dirty.password) {
        userInstance.password = await Hash.make(userInstance.password)
      }
    })
  }

  /**
   * A relationship on tokens is required for auth to
   * work. Since features like `refreshTokens` or
   * `rememberToken` will be saved inside the
   * tokens table.
   *
   * @method tokens
   *
   * @return {Object}
   */
  tokens () {
    return this.hasMany('App/Models/Token')
  }
}
```

**Reply by Mzzay** [reply]
If I log `user` I have this : 

```js
User {
  __setters__: [
    '$attributes',
    '$persisted',
    'primaryKeyValue',
    '$originalAttributes',
    '$relations',
    '$sideLoaded',
    '$parent',
    '$frozen',
    '$visible',
    '$hidden'
  ],
  '$attributes': {
    ..... 
  },
  '$persisted': true,
  '$originalAttributes': {
    ....
  },
  '$relations': {},
  '$sideLoaded': {},
  '$parent': null,
  '$frozen': false,
  '$visible': undefined,
  '$hidden': undefined
}
```

**Reply by thetutlage** [reply]
Ahh you are using v4. I thought you are on v5. 

Still the `user.password` shouldn't be undefined. Can you share the actual code you are running? Or better share a sample repo to reproduce the issue.

Also in `v4` the hash arguments are swapped. So it should be

```ts
await Hash.verify(password, user.password)
```

**Reply by Mzzay** [reply]
yes I'm on v4, sorry. 
But yes `user.password` still undefined. 
I log `user` before

**Reply by Mzzay** [reply]
Ok, it's working but I need to use `user.$attributes.password`
